examples/nlp/text_generation_transformer_decoder.py

```diff
@@ -0,0 +1,229 @@
+policy = keras.mixed_precision.Policy("mixed_float16")
+keras.mixed_precision.set_global_policy(policy)
+
+"""
+## Load data
+
+We are going to be using a wikipedia text dataset. The dataset is filtered by character
+length and parsed.
+"""
+
+# Download training data.
+keras.utils.get_file(
+  origin="https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-103-raw-v1.zip",
+  extract=True,
+)
+wiki_dir = os.path.expanduser("~/.keras/datasets/wikitext-103-raw/")
+
+# Load wikitext-103 and filter out short lines.
+wiki_train_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.train.raw")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+    .shuffle(buffer_size=256)
+)
+
+"""
+## Load Tokenizer
+Download vocabulary data to initialize WordPieceTokenizer. Preprocess training data with
+tokenizer.
+
+WordPieceTokenizer is an efficient implementation of the WordPiece algorithm used by BERT
+and other models. It will strip, lower-case and do other un-reversable preprocessing
+operations.
+"""
+
+# Download vocabulary data.
+vocab_file = keras.utils.get_file(
+  origin="https://storage.googleapis.com/tensorflow/keras-nlp/examples/bert/bert_vocab_uncased.txt",
+)
+
+# Get most frequent VOCAB_SIZE vocab tokens for faster/efficient training
+vocab_list = []
+with open(vocab_file, "r") as f:
+  for line in f:
+    vocab_list.append(line.strip())
+vocab_sublist = vocab_list[:VOCAB_SIZE]
+
+tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=vocab_sublist,
+    sequence_length=SEQ_LEN+1, # +1 for slicing train and label sequences
+)
+
+def preprocess(inputs):
+  outputs = tokenizer(inputs)
+  data = outputs[:, :-1]
+  labels = outputs[:, 1:]
+  return data, labels
+
+# Tokenize and split into train and label sequences
+ds = wiki_train_ds.map(preprocess).prefetch(tf.data.AUTOTUNE)
+
+"""
+## Model
+The model consists of the following layers:
+1. A TokenAndPositionEmbedding layer, which combines the embedding for the token and its
+position.
+2. Multiple TransformerDecoder layers, with the default causal masking. The layer has no
+cross-attention when run with decoder sequence only.
+3. Output linear layer.
+"""
+
+def create_model():
+  inputs = keras.layers.Input(shape=(SEQ_LEN,), dtype=tf.int32)
+  # embedding
+  embedding_layer = keras_nlp.layers.TokenAndPositionEmbedding(
+      vocabulary_size=VOCAB_SIZE,
+      sequence_length=SEQ_LEN,
+      embedding_dim=EMBED_DIM,
+      mask_zero=True
+  )
+  x = embedding_layer(inputs)
+  # decoder layers
+  for _ in range(NUM_LAYERS):
+    transformer_block = TransformerDecoder(
+        num_heads=NUM_HEADS,
+        intermediate_dim=FEED_FORWARD_DIM,
+    )
+    x = transformer_block(x) # no cross-attention
+  # output
+  outputs = keras.layers.Dense(VOCAB_SIZE)(x)
+  model = keras.Model(inputs=inputs, outputs=outputs)
+  loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+  model.compile(
+      "adam", loss=loss_fn,
```

Comment:
Add metrics and use a keyword argument for the optimizer

Reply:
fixed